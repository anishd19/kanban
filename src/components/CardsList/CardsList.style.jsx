import { Card } from "@mui/material";
import { styled } from "@mui/material";

export const CardsListContainer = styled(Card)(
  ({ theme }) => `
    height: 100%;
    display: flex;
    flex-direction: column;
    margin: ${theme.spacing(3)};
    padding: ${theme.spacing(3)};
    min-width: 350px;
    background-color: ${theme.palette.grey.lighter};
  `
);
