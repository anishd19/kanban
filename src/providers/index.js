import { useCards, CardsContext } from "./CardsProvider/CardsContext";
import CardsProvider from "./CardsProvider/CardsProvider";

export { useCards, CardsProvider, CardsContext };
