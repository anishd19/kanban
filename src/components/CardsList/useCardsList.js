import { useState } from "react";
import { useCards } from "providers";

const useCardsList = (column) => {
  const { removeColumnById, updateColumnTitle } = useCards();
  const [openDialog, setOpenDialog] = useState(false);

  const handleOpenDialog = () => {
    setOpenDialog(true);
  };

  const handleCloseDialog = () => {
    setOpenDialog(false);
  };

  const handleConfirmAction = () => {
    removeColumnById(column?.id);
  };

  const editTitle = (newTitle) => {
    updateColumnTitle(column.id, newTitle);
  };

  return {
    handleOpenDialog,
    handleCloseDialog,
    handleConfirmAction,
    editTitle,
    openDialog,
  };
};

export default useCardsList;
