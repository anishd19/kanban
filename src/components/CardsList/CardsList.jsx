import React from "react";
import { Droppable } from "react-beautiful-dnd";
import { Card, AddCard, EditableText, ConfirmationDialog } from "components";
import { Stack, Divider, IconButton, List } from "@mui/material";
import { Remove } from "@mui/icons-material";
import { CardsListContainer } from "./CardsList.style";
import useCardsList from "./useCardsList";

const CardsList = ({ column }) => {
  const {
    handleOpenDialog,
    handleCloseDialog,
    handleConfirmAction,
    editTitle,
    openDialog,
  } = useCardsList(column);

  return (
    <CardsListContainer>
      <Stack
        direction="row"
        alignItems="center"
        justifyContent="space-between"
        spacing={4}
      >
        <EditableText variant="h6" text={column.title} callBack={editTitle} />

        <Divider orientation="vertical" flexItem />
        <IconButton
          aria-label="delete"
          size="large"
          color="secondary"
          onClick={() => handleOpenDialog()}
          data-testid="remove-button"
        >
          <Remove />
        </IconButton>
      </Stack>
      <div
        style={{
          flexGrow: 1,
        }}
      >
        <Droppable droppableId={column.id}>
          {(provided) => (
            <div
              ref={provided.innerRef}
              {...provided.droppableProps}
              style={{ height: "100%" }}
            >
              <List>
                {column?.list?.map((listData, index) => (
                  <Card listData={listData} index={index} key={listData.id} />
                ))}
              </List>
              {provided.placeholder}
            </div>
          )}
        </Droppable>
      </div>
      <AddCard column={column} />

      <ConfirmationDialog
        open={openDialog}
        onClose={handleCloseDialog}
        onConfirm={handleConfirmAction}
        message="Are you sure you want to delete the column? All data associated with the column will be deleted."
      />
    </CardsListContainer>
  );
};

export default CardsList;
