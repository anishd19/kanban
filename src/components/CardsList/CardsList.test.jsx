import React from "react";
import { render, screen } from "@testing-library/react";
import { CardsContext } from "providers";
import { DragDropContext } from "react-beautiful-dnd";
import CardsList from "./CardsList";

const mockHandleOpenDialog = jest.fn();
const mockHandleConfirmAction = jest.fn();
const mockHandleCloseDialog = jest.fn();
jest.mock("./useCardsList", () => ({
  __esModule: true,
  default: () => ({
    handleOpenDialog: mockHandleOpenDialog,
    handleCloseDialog: mockHandleCloseDialog,
    handleConfirmAction: mockHandleConfirmAction,
    editTitle: jest.fn(),
    openDialog: false,
  }),
}));

const mockContextValue = {
  columnsData: [],
  addNewColumn: jest.fn(),
  removeColumnById: jest.fn(),
};

const onDragEnd = jest.fn();
const MockCardsProvider = ({ children }) => {
  return (
    <CardsContext.Provider value={mockContextValue}>
      <DragDropContext onDragEnd={onDragEnd}>{children}</DragDropContext>
    </CardsContext.Provider>
  );
};

describe("CardsList component", () => {
  it("renders column elements", () => {
    const column = { id: "column-1", title: "Column Title", list: [] };
    const { container } = render(
      <MockCardsProvider>
        <CardsList column={column} />
      </MockCardsProvider>
    );
    const titleElement = screen.getByText("Column Title");
    const removeButton = screen.getByLabelText("delete");
    expect(titleElement).toBeInTheDocument();
    expect(removeButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
