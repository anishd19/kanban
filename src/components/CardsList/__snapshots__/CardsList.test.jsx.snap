// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsList component renders column elements 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1wxca14-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiStack-root css-1e9ghc3-MuiStack-root"
    >
      <div
        class="MuiStack-root css-36g2gh-MuiStack-root"
      >
        <div
          class="MuiStack-root css-18ahtx1-MuiStack-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Column Title
          </h6>
          <button
            aria-label="edit"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeLarge css-1f88mc5-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
            variant="outlined"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
      />
      <button
        aria-label="delete"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeLarge css-1f88mc5-MuiButtonBase-root-MuiIconButton-root"
        data-testid="remove-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="RemoveIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13H5v-2h14v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      style="flex-grow: 1;"
    >
      <div
        data-rbd-droppable-context-id="0"
        data-rbd-droppable-id="column-1"
        style="height: 100%;"
      >
        <ul
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        />
      </div>
    </div>
    <div>
      <div
        class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Add a card
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Add a card
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add Card
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
